[MRG+2] Sphinx-gallery use

Titan-C opened the pull request:
This is a proposal to incorporate sphinx-gallery into scikit-learn. There are still some issues to work on so for know one works with the github PR of sphinx-gallery(https://github.com/sphinx-gallery/sphinx-gallery/pull/26) as the main development version.

To work on:
- [x] **Backreferences** The documented API gives at the end of every module a small gallery of the examples using that module. Examples are properly identified but the html output fails recognizing the correct path of the image and the html file. In scikit-learn's gen_rst.py, this paths are hard coded to its website structure. Sphinxgallery uses a configuration dictionary. This issue is mostly solved with  https://github.com/sphinx-gallery/sphinx-gallery/pull/26
- [x] **the sidebar hide behavior**. It just messes up with the scikit-learn design. And it is manually deactivated in the gallery implementation with the promise to fix it later:https://github.com/scikit-learn/scikit-learn/blob/master/doc/sphinxext/gen_rst.py#L494-508
- [x] **The carousel**: There is an additional image processing step for some images displayed in the carousel done in gen_rst.py, that sphinx-gallery does not perform
- [x] **Image naming**: All images generated by sphinx-gallery have a sphx_glr_ prefix. If images of the gallery are manually called by the documentation this would be broken.
- [x] More stuff I'm not aware of.
- [x] Maybe, as in nilearn, put a local copy of sphinx-gallery inside scikit-learn

amueller commented:
+1. Worked well for pystruct.
The sidebar is weird, I agree. What do you mean by "manually called bye the gallery"? Just included in the docs? Can we make the prefix configurable so that it can be dropped? Alternatively we could grep for all PNGs in the docs and change it, but why?

Titan-C commented:
> Can we make the prefix configurable so that it can be dropped? Alternatively we could grep for all PNGs in the docs and change it, but why?

The prefix was done as a workaround to keep track of the images generated by the gallery. Instead of deleting all images, as it is introduced by this hack https://github.com/scikit-learn/scikit-learn/blob/master/doc/sphinxext/gen_rst.py#L1087-L1106
Now it might be good to test it this hack continues to be necessary at all.

amueller commented:
Wait, but old images are kept, right? They are not deleted on every build iirc. If you run ``make html`` twice in a row it will keep the images from the first run.

Images are also starting with ``example_``, right? Or are they not?

Titan-C commented:
Images are not deleted at all because scikit-learn now builds in _build/html/stable and images are stored in _build/html/stable/_images. And this hack is for images in _build/html/_images, which is nowadays never populated.

images are named after the script name, no additional prefix is given.

amueller commented:
After the script name without the path? They are in the examples folder.
IIRC they still had the folder-prefix for pystruct.

Titan-C commented:
Yes. Images are named after the script name without the path. A prefix of `example_path` is only set for the sphinx label so one can reference the example with the :ref:example_.. instruction

amueller commented:
ah, thanks for the clarification

amueller commented:
Do you have an idea how we could not rename all inline images? Methinks there are a couple

lesteve commented:
> Do you have an idea how we could not rename all inline images? Methinks there are a couple

Were you stopped mid-sentence by something more important? Out of interest, what are your concerns besides PITAness (to be honest it's not that bad with the help of some scripting-fu) about renaming all the inline images?

GaelVaroquaux commented:
> Were you stopped mid-sentence by something more important? Out of interest,
> what are your concerns besides PITAness (to be honest it's not that bad with
> the help of some scripting-fu) about renaming all the inline images?

Broken links. Many people link to the docs/example.

I think that it's worth the cost, but we shouldn't be doing this too
often.

lesteve commented:
> Broken links. Many people link to the docs/example

But do they link directly to the images, something like http://scikit-learn.org/stable/_images/plot_classifier_comparison_001.png?

amueller commented:
Yeah I think that is what he means.
My concern was more about the PITA, but you are right, it is probably not that bad and someone can come up with a regexp.

Titan-C commented:
Gallery image calls are now renamed to sphinx-gallery standard prefix and suffix
The carousel image resizing is also implemented in a short sphinx extension in the conf.py.
Sidebar CSS alternative is in PR #4663 
To test this the git master branch of sphinx-gallery is needed

amueller commented:
LGTM. It would be nice if it would work with the pip version of sphinx-gallery. then I feel we don't need to make it a submodule or put a local copy anywhere, maybe just try and raise a good message on import error.

Titan-C commented:
Now that the sidebar issue is solved #4663 . PR is rebased to master. For review is the same website
http://scikit-learn.byethost7.com(May require some refresh on your browser)
I guess I caught all image renaming, and there seem not to be broken links.
I can do the sphinx-gallery release and require a pip install for scikit-learn with import error @amueller  or embed the release in scikit-learn @GaelVaroquaux

amueller commented:
I feel that building the website is not a frequent task, and we shouldn't overload the main repo.
We only need to pip install on the dev website building host and @ogrisel  / me need to have it for a release.
But I don't have a strong opinion. Maybe @ogrisel has thoughts?

GaelVaroquaux commented:
> I feel that building the website is not a frequent task, and we shouldn't
> overload the main repo.

We want our contributors to do it. I've always had the philosphy that
contributing should be made as easy as possible.

GaelVaroquaux commented:
I don't have time to review the code, but the website looks great.
Excellent work!

amueller commented:
I feel like pip-installing a pure python extension is not that much work. I agree that we want our contributors to do it, but I think it doesn't happen that often, partially because building with all plots takes so long, I think.
I feel like pip-installing a package is really not that large a burden on a contributor, but I also don't oppose adding it as a sub-module.

Btw: we require sphinx and matplotlib and latex and a latex plugin for building the docs IIRC. Why do we want to add sphinx-gallery as a submodule, but not sphinx?

GaelVaroquaux commented:
> But do they link directly to the images, something like http://scikit-learn.org
> /stable/_images/plot_classifier_comparison_001.png?

I have many blog posts that link to the images. It's somewhat standard:
if you want to show something with an image, and you use scikit-learn's
images.

I hope that I have used the explicit release number in the link rather
than the 'stable' alias.

But anyhow, I am fine for breaking this backward compat, as long as we
don't do it too often (aka: not try to do it any more).

GaelVaroquaux commented:
> Btw: we require sphinx and matplotlib and latex and a latex plugin for
> building the docs IIRC. Why do we want to add sphinx-gallery as a
> submodule, but not sphinx?

Latex is probablematic (like really), but we don't require it for the
HTML docs, neither any plugins. sphinx and matplotlib are standard and
packaged with every Python install.


It's not only about the burden of installing, it's about version
incompatibilities, and the difficulties to more forward in the
development of sphinx-gallery that this will create.

amueller commented:
I thought we needed some latex stuff for the formulas, but maybe it was some png library / plugin.

For versioning problems: ok, then lets do a submodule?

GaelVaroquaux commented:
> I thought we needed some latex stuff for the formulas,

No, it's integrated in MPL.

> For versioning problems: ok, then lets do a submodule?

Why not keeping it simple? What's the benefit of making a submodule? What
does it give us?

Technological sophistication just makes everything harder, not simpler
(planet Python is broken currently, and I haven't been able to fix it
after 2 hours of work :( ).

amueller commented:
I thought doing a submodule is the simple version? What does it give us? Being the most simple solution (two lines in git) and being able to easily sync with any version of sphinx-gallery.

GaelVaroquaux commented:
> I thought doing a submodule is the simple version?

Copying is simpler.

> What does it give us? Being the most simple solution (two lines in git)
> and being able to easily sync with any version of sphinx-gallery.

But we don't really need this. We need someone to manually update the
requirement and do a QA that everything renders well. Not an automatic
and implicit upgrade. This option pushes the QA on whoever builds the
docs. And if any kind of bug or change of API is pushed to
sphinx-gallery, sklearn would face a problem immediately.

amueller commented:
Submodules provide a manual update, not an automatic implicit one.

GaelVaroquaux commented:
> Submodules provide a manual update, not an automatic implicit one.

How does it propagate to other clone?

Would the plan then be to have a submodule and than, in the sphinx
conf.py, add it to the sys.path?

amueller commented:
the main repo stores which version of the submodule there is in a folder. It would just be in a subfolder of sklearn, so it would look like if it was copied.

Actually, I think copying might be simpler, as I think submodules need an additional command after checking out the repo (git submodule update), right @ogrisel?

agramfort commented:
Yes submodules require additional git fu such as --recursive options 

+1 for plain copy


> On 29 mai 2015, at 00:07, Andreas Mueller <notifications@github.com> wrote:
> 
> the main repo stores which version of the submodule there is in a folder. It would just be in a subfolder of sklearn, so it would look like if it was copied.
> 
> Actually, I think copying might be simpler, as I think submodules need an additional command after checking out the repo (git submodule update), right @ogrisel?
> 
> —
> Reply to this email directly or view it on GitHub.
>

Titan-C commented:
Ok, ready. A local copy of sphinx-gallery is copied from PyPi with an automatic script(also included here).

lesteve commented on doc/sphinxext/copy_sphinxgallery.sh:
```diff
@@ -0,0 +1,8 @@
+
+easy_install -Zeab tmp sphinx-gallery
+
+cp -vru tmp/sphinx-gallery/sphinxgallery/ .
```
Before I forget you will need to add the `git add` that you added in https://github.com/nilearn/nilearn/pull/592.

lesteve commented:
Note to people involved: I would wait until nilearn/nilearn#592 is merged and things settle down a bit more before looking at this PR into further details.

amueller commented:
Btw, @GaelVaroquaux latex is required for the html docs because it contains math. I just had a student try to add a formula to the docs on Os X, no fun :-/

Titan-C commented:
> Btw, @GaelVaroquaux latex is required for the html docs because it contains math. I just had a student try to add a formula to the docs on Os X, no fun :-/

Well we could migrate from using pngmath to mathjax for the docs(just changing the sphinx extension, both are default and we are not using any latex preamble for pngmath). Then you only need a working internet connection and the math is rendered by javascript.

amueller commented:
or we could have a local copy of mathjax, right?

jnothman commented:
> or we could have a local copy of mathjax, right?

I was curious about this. The mathjax zip file is 32MB, the vast majority of which is webfont data allowing cross-browser rendering. I think keep PNG default for users compiling on their own machines, but we can use MathJax on the website.

ogrisel commented:
I find it easier much for contributors (me included) to do:

    pip install sphinx sphinx-gallery

rather than mess with git submodule.

ogrisel commented:
We can also embed a local copy of the sphinx-gallery source code as in practice this is already what we do.

amueller commented:
Either would be fine with me.

Titan-C commented:
I will do a full redo of this PR once sphinx-gallery v0.0.12 is released.

Titan-C committed:
[MAINT] Install sphinx-gallery v0.1.0

Titan-C committed:
conf.py integrate sphinx-gallery

Titan-C committed:
update image sources namespace

Titan-C committed:
update references link namespace

Titan-C committed:
update reference namespace in gallery scripts

Titan-C committed:
Add carousel

Fix reference link

Titan-C committed:
References to examples

Titan-C commented:
PR redone to incorporate Sphinx-Gallery v0.1.0
It looks really nice, and I hope not to have missed any image, reference or link rename

Titan-C committed:
more examples references

Titan-C commented:
As there is no CircleCI to build the documentation. I have uploaded my build for review at http://scikit-learn.byethost7.com/

lesteve commented:
Weird your generated documentation webpage shows the stable version as 0.16. Can you double-check that your link is correct?

Titan-C commented:
> Weird your generated documentation webpage shows the stable version as
> 0.16. Can you double-check that your link is correct?
I have experienced a lot of forced caching in this web domain. So force a
reload, because to me it shows version 0.18.dev0

lesteve commented:
> I have experienced a lot of forced caching in this web domain. So force a reload, because to me it shows version 0.18.dev0

Still the same. Here is the snapshot for clarity.
![sphinx-gallery](https://cloud.githubusercontent.com/assets/1680079/12551796/62a1203c-c36d-11e5-9310-62b694718bf2.png)

Also I noticed a problem, it looks like the `make html-noplot` does not work anymore, i.e. the examples are always run. It seems like this is a sphinx-gallery problem AFAICT.

lesteve commented:
> Still the same. Here is the snapshot for clarity.

Actually I am not sure what you did but:
* http://scikit-learn.byethost7.com points to some outdated webpage (0.17 is the development version)
* http://scikit-learn.byethost7.com/auto_examples/index.html points to (I am guessing) your locally generated webpage although it should point to the examples gallery unless I am missing something.
* http://scikit-learn.byethost7.com/auto_examples/auto_examples/index.html points to the gallery (two auto_examples instead of one)

Can you figure out what is happening and clean it up?

Titan-C commented:
> Can you figure out what is happening and clean it up?

website should be clean now.

lesteve commented:
> website should be clean now.

Looks better now, thanks !

GaelVaroquaux commented:
> As there is no CircleCI to build the documentation. I have uploaded my build
> for review at http://scikit-learn.byethost7.com/

It looks like it uncovered a glitch in sphinx-gallery:
http://scikit-learn.byethost7.com/auto_examples/applications/plot_out_of_core_classification.html

This example is not formatted accordingly to the notebook-style of
sphinx-gallery. However, the behavior of sphinx-gallery on this example
is arguably a bug: it seems to remove the "#" sign in some comments.

Titan-C committed:
[MAINT] Update to sphinx-gallery v0.1.1

Titan-C committed:
Fix example to use rst syntax in notebook comments

Titan-C commented:
I rebased to master and updated to sphinx-gallery v0.1.1 that fixes the CSS and the no-plot bug.
Regarding the example showing the ill behavior of splitting the notebook styled file wrong. With @lesteve  we agreed to change the example, as it is the only one to use that syntax. A comment block for a script that is going to be parsed by sphinx-gallery has rst syntax. And comments according to  pep 8 have a space after the #. 
so the comment 
```
#########################
# Main
#########################
```
Must be
```
#########################
# Main
# #######################
```
For text and an horizontal line, and be properly rendered by the gallery
or it can take advantage of the rst features and be
```
#########################
# Main
# ----
```
which is the current choice, to make those breaks headers.

The new build is here http://scikit-learn.byethost7.com/stable/index.html

You will also notice that some examples are now broken. This is because I used scipy 0.17 where `The function misc.lena has been removed due to license incompatibility.` What should I do in this case?

GaelVaroquaux commented:
<pre>
#########################
# Main
#########################
</pre>

That's a very natural formatting to use, and many people use it, inside and outside scikit-learn. It seems quite natural that such a formatting would yield a title.

Is it possible to support it, without super strange rules?

GaelVaroquaux commented:
I forgot to mention: for the scikit-learn devs that might not be aware: the big deal of this updated version of sphinx-gallery is that it generates IPython notebooks from the examples. Splitting cells requires a specific syntax (namely lines with many comment hashes, as in "#####################"), but it's a natural syntax that we use already in several examples.

Titan-C commented:
> Is it possible to support it, without super strange rules?

Not directly, Sphinx-Gallery would need an extra parser for the text
blocks to catch this one. Maybe doing it 2 pass, like the IPython
notebook parser does to work with equations. I can open a new issue in
Sphinx-Gallery to try to support this. But for this release I would set
the only example using this syntax to have rst syntax.

lesteve commented:
> ```
#########################
# Main
#########################
```
> That's a very natural formatting to use, and many people use it, inside and outside scikit-learn. It seems quite natural that such a formatting would yield a title.

I would need to be convinced about this one. For simplicity's sake I'd rather only support rst inside a text cell. For example what should happen if we have something like:

```
#########################
# Main
#########################
# Some more text
```

Is that two text cells (which it is at the moment I think)? Is that one title followed by a comment that should go in a code cell?

Just as a reminder this syntax is only used in one scikit-learn example, MNE-python doesn't seem to use it otherwise they would have complained already. Maybe we should look in other projects to see how much it is used.

I don't remember the exact reason why the rendering was so awkward. A solution would be to tweak the regex pattern so that the rendering is slightly more acceptable, just a wild suggestion, haven't looked into it in more details.

Titan-C commented:
Sphinx Gallery is rebased to master.

Titan-C committed:
Extra missing links complained by sphinx

GaelVaroquaux commented:
@Titan-C : can you include the latest sphinx-gallery in this PR.

Once this is done, I think that we should merge: it not only will fix #6310, but also give us a route forward to fixing other similar problems.

Titan-C commented:
Sphinx-Gallery v0.1.2 released and updated into this PR

Titan-C commented:
Sphinx-Gallery v0.1.2 released and included in this PR

Titan-C committed:
[MAINT] update to sphinx-gallery v0.1.2

Titan-C committed:
Remove custom touch to backreferences files

This is now a default in sphinx-gallery v0.1.2

GaelVaroquaux commented:
CircleCi is failing. There are a lot of warnings and errors. One of them is:
`
/home/ubuntu/scikit-learn/doc/modules/clustering.rst:607: WARNING: undefined label: sphx_glr_example_cluster_plot_face_ward_segmentation.py (if the link has no caption the label must precede a section header)
`

But there are many more.

Titan-C committed:
Sphinx-Gallery has to be loaded after autodoc

Sphinx loads its extensions sequentially as they are specified on the
list. Sphinx-Gallery relies en autodoc to generated the example
thumbnails when auto documenting the API. If autodoc is not loaded
sphinx-gallery skips this step, as it implies the user might not want
it. In this case we need it, so the gallery has to be loaded after
autodoc and autosummary.

The autosummary_generate variable is repeated

Titan-C commented:
I think this time is working and I'm not missing more links.

Titan-C committed:
Python 2 backwards compatibility in example

This lines breaks the gallery build in python 2, which is the version on
which CircleCI runs. Other option is to build the docs on Python 3

Titan-C commented:
I found the failing example. I was mislead by all the fail messages of sphinx and thought those where the reason for the fail in Circle CI. As I can build the gallery without any problem in my computer. I have changed the example to work on python2 which is the interpreter on CircleCi. The error was the print function having the flush keyword.
Or would you prefer to have python3 from conda as the default interpreter? Change the build_doc.sh script to it?

lesteve commented:
@Titan-C you need to rebase on master and fix conflicts.

lesteve commented:
> I have changed the example to work on python2 which is the interpreter on CircleCi. The error was the print function having the flush keyword.
Or would you prefer to have python3 from conda as the default interpreter? Change the build_doc.sh script to it?

Changing the example was the way to go. Examples should be runnable with both python 2 and 3.

Titan-C commented:
Rebased

lesteve commented:
Looks like Travis is failing for a reason that is unrelated to the PR.

lesteve commented:
Maybe do a git commit --amend and force push so that everything is green.

Titan-C commented:
> Maybe do a git commit --amend and force push so that everything is
> green.

What shall I amend? Problem is from conda MKL and master is failing in
similar way too.

lesteve commented:
> What shall I amend? Problem is from conda MKL and master is failing in
similar way too.

I think you are right, leave it like that for now. `git commit --amend` + force push is a work-around to get all the CIs to rerun. It's useful from time to time when there are temporary glitches.

amueller commented:
squash?

Not sure if the jupyter cell formatting has been resolved but I had a silly question:
Shouldn't jupyter notebook examples have a ``.ipynb`` ending, and pure python examples have a ``.py`` ending? I wouldn't expect something in a ``.py`` file to be parsed as notebook.

I don't have a strong opinion on changing the examples for now, but this seems weird to me.

Titan-C commented:
> squash?
> 
> Not sure if the jupyter cell formatting has been resolved but I had a silly question:
> Shouldn't jupyter notebook examples have a .ipynb ending, and pure python examples have a .py ending? I wouldn't expect something in a .py file to be parsed as notebook.
> 
> I don't have a strong opinion on changing the examples for now, but this seems weird to me.

I don't understand what you are talking about. The gallery of examples is written in plain text python files, indifferent of having just the docstring and the code or a richer commenting style to be parsed by blocks into a notebook styled example. There are no Jupyter Notebook examples as source. Sphinx gallery copies the original python text file to make it available for download. Then parses it into RST, then to HTML for the website and in parallel parses it into a Jupyter Notebook and makes it available for download too.

Titan-C commented:
Rebased to master

amueller commented:
broken again :-/

amueller commented:
btw, I really think there is no reason to add this as a subpackage. We should really start properly handling our dependencies now that numpy and scipy are properly pip installable.

Titan-C commented:
It's rebased to master.

> btw, I really think there is no reason to add this as a subpackage. We should really start properly handling our dependencies now that numpy and scipy are properly pip installable.

I'm fine with that too. How is that you are building and publishing documentation now?
I also should release a new version of sphinx-gallery in the coming week, it should give advise on the failing examples.

amueller commented:
I'm good with merging this as-is for now. @GaelVaroquaux @agramfort ?

amueller commented:
I'd prefer this to be a dependency, but that is a separate discussion. We should avoid breaking this PR again.

agramfort commented:
there are plenty of broken image links. See eg:

https://4184-843222-gh.circle-artifacts.com/0//home/ubuntu/scikit-learn/doc/_build/html/stable/modules/linear_model.html

also in the carrousel etc. this needs more careful edits to be sure we don't break the website.

regarding dependency no strong feeling

GaelVaroquaux commented:
> regarding dependency no strong feeling

I've just had a somewhat painful update of Ubuntu which broke too many
things. So I guess that I am still very much in favor of static linking
(ie embedding source code, in Python) rather than dependencies.

I'll review this PR once @Titan-C fixes the broken images.

lesteve commented:
> there are plenty of broken image links. See eg:

> https://4184-843222-gh.circle-artifacts.com/0//home/ubuntu/scikit-learn/doc/_build/html/stable/modules/linear_model.html

FWIW I couldn't spot any broken images on this page but maybe I am missing something.

The carousel does have some broken images though (7th and 10th image).

If I understood correctly @Titan-C want to do a sphinx-gallery release before updating this PR. There may be a few loose ends to tie before doing the sphinx-gallery release.

agramfort commented:
see
https://4184-843222-gh.circle-artifacts.com/0//home/ubuntu/scikit-learn/doc/_build/html/stable/modules/linear_model.html#huber-regression

image plot_huber_vs_ridge_001.png

Titan-C commented:
I just rebased on master, but master is not passing on travis
I fixed the carousel images and the one in linear model
I'll keep looking in my local build log if something else is missing
@lesteve @GaelVaroquaux @agramfort @amueller @ogrisel need some eyes

amueller commented:
@GaelVaroquaux Maybe we can discuss dependency management at some point. I feel that pipy is in a relatively good place, and I don't really see any issue with reasonable dependencies and them being resolved by the package manager. Also, we seem to be the only package that is doing that, and others don't seem to run into trouble.

amueller commented:
master should be passing soon, just cleared the cache.

GaelVaroquaux commented:
> @GaelVaroquaux Maybe we can discuss dependency management at some point. I feel
> that pipy is in a relatively good place, and I don't really see any issue with
> reasonable dependencies and them being resolved by the package manager.

It's a QA problem: there is a huge combinatory, and embedding packages
reduces it. It's not a packaging problem per se: that one has improved
a lot.

> Also, we seem to be the only package that is doing that, and others
> don't seem to run into trouble.

That's a question of point of view. My feeling is that the state of
PyData is quite bad and that minor things are often broken.

Titan-C committed:
Sphinx-doc absolute path for auto_example images

agramfort commented:
no circleci ?

GaelVaroquaux commented:
Indeed, it seems that github's linking to circleCI is broken. The circleCI url of the PR is https://circleci.com/gh/scikit-learn/scikit-learn/tree/pull%2F4658

agramfort commented:
LGTM

+1 for merge

thanks heaps @Titan-C

Titan-C commented:
This PR is always failing in travis. I rebased to master which is now passing, but it still fails. I don't understand why this change in the docs results in failing the tests.

https://travis-ci.org/Titan-C/scikit-learn/builds/148056947

amueller commented:
Can you please push again? I messed up CI. Sorry.

Titan-C commented:
I have no idea how to pass the tests. Master passes on travis. On my computer it fails with import error
```
ERROR: Failure: ImportError (bad magic number in 'sklearn._build_utils.cythonize
```
The sphinx-gallery branch fails on travis too with 
```
ImportError: cannot import name predict_stages
```

GaelVaroquaux commented:
@Titan-C : this looks great, and I really want to merge it. But indeed, travis has a spurious failure.

Can you rebase on master and force push to see if that fixes the failure.

lesteve commented:
I cleared the Travis cache for this PR and restarted Travis. This has previously solved similar problems in other PRs, let's see about this one.

lesteve commented:
@Titan-C is the sphinx-gallery copy from a sphinx-gallery release or from master?

Titan-C commented:
I'm rebasing now for a forced push again.

> is the sphinx-gallery copy from a sphinx-gallery release or from master?

It downloads from pip

GaelVaroquaux commented:
This is awesome. I checked on circle-CI and on my box, everything is fine. Merging. Thanks heaps @Titan-C !

amueller commented:
awesome, thanks!

amueller commented:
@GaelVaroquaux btw, building the doc also requires
> dvipng texlive-latex-base texlive-latex-extra
In addition to
> python numpy scipy cython nose coverage matplotlib sphinx pillow

according to the "build_doc.sh" circle_ci script. Not sure how nose and coverage got in there, but the rest is certainly mandatory. I feel this slightly violates the "no dependency" policy ;)

GaelVaroquaux commented:
>     dvipng texlive-latex-base texlive-latex-extra
>     In addition to
>     python numpy scipy cython nose coverage matplotlib sphinx pillow

> according to the "build_doc.sh" circle_ci script. Not sure how nose and
> coverage got in there, but the rest is certainly mandatory. I feel this
> slightly violates the "no dependency" policy ;)

I am actually not super excited about the dvipng and related TeX
dependencies.

The way that I think about it is that sphinx-gallery is still a young
project that evolves fast. As it becomes more mature, we could consider
moving it to a dependency for the docs, as the docs can have more
dependencies that the run-time dependencies.

amueller commented:
:+1: seems reasonable.
